File: oolab/src/test/java/agh/ics/oop/model/Vector2dTest.java
package agh.ics.oop.model;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class Vector2dTest {
    @Test
    public void equalsTest() {
        var v1 = new Vector2d(1, 2);
        var v2 = new Vector2d(2, 2);
        var v3 = new Vector2d(1, 2);
        assertEquals(v1, v3);
        assertEquals(v3, v1);
        assertNotEquals(v1, v2);
        assertNotEquals(v2, v1);
    }

    @Test
    public void toStringTest() {
        var v = new Vector2d(1, 2);
        assertEquals(v.toString(), "(1,2)");
    }

    @Test
    public void precedesTest() {
        var v1 = new Vector2d(0, 0);
        var v2 = new Vector2d(2, 0);
        var v3 = new Vector2d(2, 2);
        var v4 = new Vector2d(0, 2);
        assertTrue(v1.precedes(v1));
        assertTrue(v1.precedes(v2));
        assertTrue(v1.precedes(v3));
        assertTrue(v1.precedes(v4));
        assertTrue(v2.precedes(v2));
        assertTrue(v2.precedes(v3));
        assertTrue(v4.precedes(v3));
        assertTrue(v4.precedes(v4));
    }

    @Test
    public void followsTest() {
        var v1 = new Vector2d(0, 0);
        var v2 = new Vector2d(2, 0);
        var v3 = new Vector2d(2, 2);
        var v4 = new Vector2d(0, 2);
        assertTrue(v1.follows(v1));
        assertTrue(v2.follows(v1));
        assertTrue(v2.follows(v2));
        assertTrue(v3.follows(v1));
        assertTrue(v3.follows(v2));
        assertTrue(v3.follows(v4));
        assertTrue(v4.follows(v1));
        assertTrue(v4.follows(v4));
    }

    @Test
    public void upperRightTest() {
        var v1 = new Vector2d(0, 0);
        var v2 = new Vector2d(2, 0);
        var v3 = new Vector2d(2, 2);
        var v4 = new Vector2d(0, 2);
        assertEquals(v1.upperRight(v2), v2);
        assertEquals(v1.upperRight(v4), v4);
        assertEquals(v2.upperRight(v4), v3);
    }

    @Test
    public void lowerLeftTest() {
        var v1 = new Vector2d(0, 0);
        var v2 = new Vector2d(2, 0);
        var v3 = new Vector2d(2, 2);
        var v4 = new Vector2d(0, 2);
        assertEquals(v1.lowerLeft(v2), v1);
        assertEquals(v3.lowerLeft(v4), v4);
        assertEquals(v2.lowerLeft(v4), v1);
    }

    @Test
    public void addTest() {
        var v1 = new Vector2d(1, 2);
        var v2 = new Vector2d(-2, -1);
        var v3 = new Vector2d(-1, 1);
        assertEquals(v1.add(v2), v3);
    }

    @Test
    public void subtractTest() {
        var v1 = new Vector2d(1, 2);
        var v2 = new Vector2d(-2, -1);
        var v3 = new Vector2d(3, 3);
        assertEquals(v1.subtract(v2), v3);
    }

    @Test
    public void oppositeTest() {
        var v1 = new Vector2d(1, 2);
        var v2 = new Vector2d(-1, -2);
        assertEquals(v1.opposite(), v2);
        assertEquals(v1.opposite().opposite(), v1);
    }
}

File: kotlinlab/src/test/kotlin/model/Vector2dTest.kt
package model

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe

class Vector2dTest : FunSpec({

    test("equals") {
        val v1 = Vector2d(1, 2)
        val v2 = Vector2d(2, 2)
        val v3 = Vector2d(1, 2)
        v1 shouldBe v3
        v3 shouldBe v1
        v1 shouldNotBe v2
        v2 shouldNotBe v1
    }

    test("toString") {
        Vector2d(1, 2).toString() shouldBe "(1,2)"
    }

    test("precedes") {
        val v1 = Vector2d(0, 0)
        val v2 = Vector2d(2, 0)
        val v3 = Vector2d(2, 2)
        val v4 = Vector2d(0, 2)
        v1.precedes(v1) shouldBe true
        v1.precedes(v2) shouldBe true
        v1.precedes(v3) shouldBe true
        v1.precedes(v4) shouldBe true
        v2.precedes(v2) shouldBe true
        v2.precedes(v3) shouldBe true
        v4.precedes(v3) shouldBe true
        v4.precedes(v4) shouldBe true
    }

    test("follows") {
        val v1 = Vector2d(0, 0)
        val v2 = Vector2d(2, 0)
        val v3 = Vector2d(2, 2)
        val v4 = Vector2d(0, 2)
        v1.follows(v1) shouldBe true
        v2.follows(v1) shouldBe true
        v2.follows(v2) shouldBe true
        v3.follows(v1) shouldBe true
        v3.follows(v2) shouldBe true
        v3.follows(v4) shouldBe true
        v4.follows(v1) shouldBe true
        v4.follows(v4) shouldBe true
    }

    test("upperRight") {
        val v1 = Vector2d(0, 0)
        val v2 = Vector2d(2, 0)
        val v3 = Vector2d(2, 2)
        val v4 = Vector2d(0, 2)
        v1.upperRight(v2) shouldBe v2
        v1.upperRight(v4) shouldBe v4
        v2.upperRight(v4) shouldBe v3
    }

    test("lowerLeft") {
        val v1 = Vector2d(0, 0)
        val v2 = Vector2d(2, 0)
        val v3 = Vector2d(2, 2)
        val v4 = Vector2d(0, 2)
        v1.lowerLeft(v2) shouldBe v1
        v3.lowerLeft(v4) shouldBe v4
        v2.lowerLeft(v4) shouldBe v1
    }

    test("plus") {
        Vector2d(1, 2) + Vector2d(-2, -1) shouldBe Vector2d(-1, 1)
    }

    test("minus") {
        Vector2d(1, 2) - Vector2d(-2, -1) shouldBe Vector2d(3, 3)
    }

    test("opposite") {
        val v1 = Vector2d(1, 2)
        val v2 = Vector2d(-1, -2)
        -v1 shouldBe v2
        -(-v1) shouldBe v1
    }
})